Fix up handling of environment variables

rhatdan opened the pull request:
The way docker works is if a user specifies a non `-e Name=Value`, IE
just a `-e Name`, then the environment variable Name from the clients
OS.ENV is used.

Also by default Docker containers run with the HOSTNAME environment set
to the HOSTNAME specified for the container.

Signed-off-by: Daniel J Walsh <dwalsh@redhat.com>

rhatdan commented:
@baude @mheon @umohnani8 PTAL

mheon commented on cmd/kpod/create_cli.go:
```diff
@@ -26,25 +26,6 @@ func getAllLabels(labelFile, inputLabels []string) (map[string]string, error) {
 	return labels, nil
 }
 
-func getAllEnvironmentVariables(envFiles, envInput []string) ([]string, error) {
```
Curious as to why you're breaking this up. It doesn't seem necessary to completely refactor it out - can we just pass in a spec generator and not return anything?

rhatdan replied:
I am not breaking anything up?  I removed a non useful function, that was throwing errors on 
`--env foo`, commands, If you want me to make a separate function for the spec file generation, I can do that.

mheon replied:
Alright, fair enough. Would be difficult to test anyways, given that we'd only be modifying the spec generator.

rhatdan replied:
@mheon These three lines replace the getAllEnvironmentVariables.

```
-       env, err := getAllEnvironmentVariables(c.StringSlice("env-file"), c.StringSlice("env"))
+       env, err := readKVStrings(c.StringSlice("env-file"), c.StringSlice("env"))
        if err != nil {
                return &createConfig{}, errors.Wrapf(err, "unable to process environment variables")
        }
+       env = append(defaultEnvVariables, env...)
```

TomSweeneyRedHat commented on cmd/kpod/spec.go:
```diff
@@ -124,6 +128,22 @@ func createConfigToOCISpec(config *createConfig) (*spec.Spec, error) {
 		g.AddTmpfsMount(spliti[0], options)
 	}
 
+	for _, i := range config.env {
```
Did you decide to not append the defaultEnvVariables if none were found?  i.e. line 34 from the deleted section of create_cli.go above.

rhatdan replied:
We are always setting the default environment above in the patch.

+       env = append(defaultEnvVariables, env...)

mheon replied:
I don't think the check is necessary, given that appending an empty array is a no-op

TomSweeneyRedHat replied:
Missed that, thx, I'll blame the cold meds.

TomSweeneyRedHat commented on test/kpod_run.bats:
```diff
@@ -56,3 +56,23 @@ ALPINE="docker.io/library/alpine:latest"
+
+@test "run environment test" {
+
+#    run bash -c "${KPOD_BINARY} ${KPOD_OPTIONS} run ${ALPINE} sh -c printenv | grep HOSTNAME"
```
Curious why this is commented out?

rhatdan replied:
We don't set the hostname in our containers since we are not using networks yet.  As soon as we have network support for kpod run, we can uncomment this.  I will add a comment on this.

mheon commented on cmd/kpod/create.go:
```diff
@@ -264,10 +264,11 @@ func parseCreateOpts(c *cli.Context, runtime *libpod.Runtime) (*createConfig, er
 	if err != nil {
 		return &createConfig{}, errors.Wrapf(err, "unable to process environment variables")
 	}
+	env = append(defaultEnvVariables, env...)
```
Nit: this seems a little backward - usually would do append(env, defaultEnvVariables...) to make it more clear. Doesn't actually effect the results just a style preference

rhatdan replied:
I did it this way so that if path is defined in defaultEnvVariables and in --env path, then the second path will replace the first path.

If I reverse then the user would not be able to override the defaults.

mheon commented:
One nit, otherwise LGTM. Though I am also curious why part of the environment tests is commented out.

rhatdan commented:
I reworked the function to setupEnvironment in spec.go, added a comment about the HOSTNAME check being commented out.
Added a test to make sure PATH= is overriden defaultEnvironment.

TomSweeneyRedHat commented:
@rhatdan looks like you're getting bit by some travis environment bogosity.

rhatdan commented:
bot retest

rhatdan commented:
bot retry please

rhatdan commented:
@jlebon @cgwalters Is there a way to keep the test image around, so we can log in and take a look around.
We are getting strange behaviour on the Centos test AMI,  we think there could be a mismatch or some issue with the runc being used on this box.

We can not seem to get this failure on the centos machine we run and the error does not happen in travis or in the Fedora test machine.

The error we are seeing happens when the runc RuntimeSpec version does not match the kpod RunTime Spec version, and yet the rpm values look correct.

jlebon commented:
Not currently, though you should be able to reproduce the same environment by starting from the same image it uses. But maybe first, let's move away from `centos/7/atomic/alpha`? It's no longer updated and the image there is pretty old 🕸 by now. Try `centos/7/atomic/smoketested` ✨ instead!

The corresponding image can be downloaded here to test locally: https://ci.centos.org/artifacts/sig-atomic/centos-continuous/images-smoketested/cloud/latest/images/.

rhatdan commented:
Ok trying with smoketested

rhatdan commented:
@baude @mheon PTAL

baude commented:
@rh-atomic-bot r+ da86799
